validate-deployment-readiness

```diff
@@ -1,77 +1,106 @@
-      tr -d '\n' | # Remove trailing newline
-      awk -F / '{print $NF}' # Filter down to the last field after a '/' character,
-                             # e.g. foo/bar/x.js becomes x.js
-  )
-
-  local_hash=$(echo -n $latest_local_migration | shasum -a 256 | cut -d ' ' -f 1)
-  echo $c_yellow'Local hash: '$c_reset$local_hash
-
-  if [[ $deployed_hash != $local_hash ]] ; then
-    echo $c_red'Hash mismatch! Make sure to run migrations _before_ deploying. Aborting.'$c_reset
-    exit 1
-  fi
-
-  echo $c_green'Migration hashes match. Ready to roll.'$c_reset
+#!/usr/bin/env node
+
+const process = require('process');
+const https = require('https');
+const fs = require('fs');
+const crypto = require('crypto');
+
+const green = '\033[0;32m';
+const red = '\033[0;31m';
+const yellow = '\033[1;33m';
+const reset = '\033[0m';
+
+const hash = crypto.createHash('sha256');
+
+main()
+  .then(() => process.exit(0))
+  .catch((err) => {
+    console.error(err);
+    process.exit(1);
+  });
+
+async function main() {
+  const [_, __, target, migrationPath] = process.argv;
+
+  if (!target || !migrationPath) {
+    console.log('Usage: validate-deployment-readiness [staging|production] [path_to_migrations]');
+    process.exit(1);
+  }
+
+  let deployedHash;
+  try {
+    deployedHash = await getDeployedHash(target);
+  } catch (error) {
+    console.log(`${yellow}Warning: Bad response status:${reset}\n${JSON.stringify(error, null, 2)}`);
+    process.exit(0);
+  }
+  console.log(`${yellow}Deployed hash: ${reset}${deployedHash}`);
+
+  const latestLocalMigration = await getLocalMigration(migrationPath);
+  const localHash = hash.update(latestLocalMigration).digest('hex');
+  console.log(`${yellow}Local hash:    ${reset}${localHash} (from file: ${latestLocalMigration})`);
+
+  if (deployedHash === localHash) {
+    console.log(`${green}Migration hashes match. Ready to roll.${reset}`);
+    process.exit(0);
+  } else {
+    console.error(`${red}Hash mismatch! Make sure to run migrations _before_ deploying. Aborting.${reset}`);
+    process.exit(1);
+  }
 }
 
-main "$@"
+function getDeployedHash(target) {
+  return new Promise((resolve, reject) => {
+    const hostname = target === 'staging' ? 'api-stg.ca.la' : 'api.ca.la';
+
+    let response = [];
+    const req = https.request(
+      {
+        hostname,
+        port: 443,
+        path: '',
+        method: 'GET'
+      },
+      res => {
+        res.on('data', d => response = response.concat(d));
+        res.on('end', () => {
+          const responseJson = JSON.parse(response.join(''));
+
+          if (responseJson.status === 'ok') {
+            resolve(responseJson.lastMigrationHash);
+          } else {
+            reject(reponseJson);
+          }
+        })
+      }
+    )
+
+    req.on('error', reject);
+    req.end();
+  });
+}
+
+function getLocalMigration(migrationPath) {
+  return new Promise((resolve, reject) => {
+    fs.readdir(migrationPath, (err, files) => {
```

Comment:
Just to double-check my own understanding here: `readdir` will resolve this path relative to the working directory this script is run in, not the location of the script itself, right?

Reply:
Good question. I was testing by passing an absolute path. I'll check it out and make sure it does the right thing.

Reply:
Read up a bit more, I'm pretty sure this is correct as-is: https://nodejs.org/api/fs.html#fs_file_paths

This is why you'll see `__dirname` so often when working with paths in node: to explicitly prevent this behavior of resolving relative to CWD.

Reply:
This seems to work as expected:

```
../bin/validate-deployment-readiness staging dist/migrations
Deployed hash: 53b39f5002e4d809adf1797f90b969cf704a37f7ccffbab293f7079c08397aa4
Local hash:    53b39f5002e4d809adf1797f90b969cf704a37f7ccffbab293f7079c08397aa4 (from file: 20180727143038_restrict-design-to-single-collection.js)
Migration hashes match. Ready to roll.
```